Src/WindowsServer/WindowsServer.Shared/Implementation/AzureRoleEnvironmentContextReader.cs

```diff
@@ -62,25 +86,76 @@ internal set
+
+        internal static string Culture { get; set; }
+
+        internal static string PublicKeyToken { get; set; }
+
+        internal static string[] VersionsToAttempt { get; set; }
+
         /// <summary>
         /// Initializes the current reader with respect to its environment.
         /// </summary>
         public void Initialize()
         {
-            ServiceRuntime serviceRuntime = new ServiceRuntime();
-            RoleEnvironment roleEnvironment = serviceRuntime.GetRoleEnvironment(AzureRoleEnvironmentContextReader.BaseDirectory);
+            AppDomain tempDomainToLoadAssembly = null;
+            string tempDomainName = "AppInsightsDomain-" + Guid.NewGuid().ToString().Substring(0, 6);
+            long beginTimeInTicks = Stopwatch.GetTimestamp();
 
-            if (roleEnvironment.IsAvailable == true)
+            // The following approach is used to load Microsoft.WindowsAzure.ServiceRuntime assembly and read the required information.
+            // Create a new AppDomain and try to load the ServiceRuntime dll into it.
+            // Then using reflection, read and save all the properties we care about and unload the new AppDomain.            
+            // This approach ensures that if the app is running in Azure Cloud Service, we read the necessary information deterministically
+            // and without interfering with any customer code which could be loading same/different version of Microsoft.WindowsAzure.ServiceRuntime.dll.
+            try
             {
-                RoleInstance roleInstance = roleEnvironment.CurrentRoleInstance;
-                if (roleInstance != null)
+                AppDomainSetup domaininfo = new AppDomainSetup();            
+                domaininfo.ApplicationBase = AzureRoleEnvironmentContextReader.BaseDirectory;                
+
+                // Create a new AppDomain
+                tempDomainToLoadAssembly = AppDomain.CreateDomain(tempDomainName, null, domaininfo);
+
+                // Load the RemoteWorker assembly to the new domain            
+                tempDomainToLoadAssembly.Load(typeof(AssemblyLoader).Assembly.FullName);
+
+                // Any method invoked on this object will be executed in the newly created AppDomain.
+                AssemblyLoader remoteWorker = (AssemblyLoader)tempDomainToLoadAssembly.CreateInstanceAndUnwrap(typeof(AssemblyLoader).Assembly.FullName, typeof(AssemblyLoader).FullName);
+                remoteWorker.AssemblyName = AzureRoleEnvironmentContextReader.AssemblyName;
+                remoteWorker.Culture = AzureRoleEnvironmentContextReader.Culture;
+                remoteWorker.PublicKeyToken = AzureRoleEnvironmentContextReader.PublicKeyToken;
+                remoteWorker.VersionsToAttempt = AzureRoleEnvironmentContextReader.VersionsToAttempt;
+                bool success = remoteWorker.ReadAndPopulateContextInformation(ref this.roleName, ref this.roleInstanceName);
```

Comment:
you need to check whether it's emulator somewhere. In Emulator we may need to implement different behavior. Depending on what emulator returns as RoleName and RoleInstance we may want to not use provided values